rooms.js

```diff
@@ -990,7 +990,9 @@ class BattleRoom extends Room {
 	update(excludeUser) {
 		if (this.log.length <= this.lastUpdate) return;
 
-		Sockets.subchannelBroadcast(this.id, '>' + this.id + '\n\n' + this.log.slice(this.lastUpdate).join('\n'));
+		if (this.userCount) {
```

Comment:
I thought this was an optimization, but tests actually fail without the ``if`` statement:

> Attempted to broadcast to subchannels of nonexistent channel battle-gen7customgame-2329

Something is seriously wrong in how we handle joining rooms...

Reply:
No, that seems correct to me?

Reply:
Like, a channel isn't exactly a room, empty channels are always auto-deleted.

Reply:
Morfent has a distaste for broadcasting to nonexistent channels and considers it a bug, but I always designed the channel system to consider "nonexistent channel" and "empty channel" the same thing.

Reply:
For context, the mentioned assertion failure happens when p1 tries to join a battle room for the first time, and the room sends the ``|j|[P1IDENTITY]`` message.

> Morfent has a distaste for broadcasting to nonexistent channels and considers it a bug

And I agree with him.

There should have been a ``Sockets.channelAdd()`` call before trying to send messages to this room, rather than relying on ``.userCount``, which is inherently undefined on ``user.joinRoom()``.

Yes, it all actually works fine together; however, it's extremely confusing behavior, as well as highly error-prone when implementing new features, such as [shadowjoins (rejected on policy rather than technical reasons)](https://github.com/Zarel/Pokemon-Showdown/pull/2066)

Reply:
I feel like having extra state that does nothing useful is bad, though.